### PR TITLE
Non-appliaction endpoints coverage

### PR DESCRIPTION
**Pull request by pjgg**
This PR cover non-application endpoint behavior:

* Relative paths: 

```
# Application.properties example
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=q
quarkus.http.redirect-to-non-application-root-path=false
```

Valid request example: `http://localhost:8080/api/q/health`

* Empty non-application root path:

```
# Application.properties example
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=/
quarkus.http.redirect-to-non-application-root-path=false
```
All request will return an HTTP status `404`

* Non-application root path with `slash` base path:
```
# Application.properties example
%emptyRootPath.quarkus.http.root-path=/
%emptyRootPath.quarkus.http.non-application-root-path=/q
%emptyRootPath.quarkus.http.redirect-to-non-application-root-path=false
```

Valid request example: `http://localhost:8080/q/health`

* Defined application base path and non-application root path:

```
# Application.properties example
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=/q
quarkus.http.redirect-to-non-application-root-path=false
```

Valid request example: `http://localhost:8080/q/health`

* Backward compatibility (allow redirections):

```
quarkus.http.root-path=/api
quarkus.http.non-application-root-path=/q
quarkus.http.redirect-to-non-application-root-path=true
```

Valid request example: `http://localhost:8080/api/health`, will be redirected to `http://localhost:8080/q/health`

**Note:** If you want to run this branch on your local environment you will need to download and compile the following [branch](https://github.com/kenfinnigan/quarkus/tree/non-app-endpoints)

### CODE REVIEW

**Comment by pjgg** [comment]
Do not merge till the PR is in

File: 020-quarkus-http-non-application-endpoints/README.md
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/README.md:8]: There is an extra space in the above two lines.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java:18]: Have you considered using QuarkusProdModeTest like in https://github.com/quarkus-qe/beefy-scenarios/blob/901bd56412e26a79a0fbd7223f577f0ab372960c/019-quarkus-consul/src/test/java/io/quarkus/qe/GreetingResourceTest.java#L33 ?

Saying this because `@TestProfile` is only supported in JVM whereas using QuarkusProdModeTest supports both JVM and Native, also we can specify different properties by test case here.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java:24]: I would use a parametrise test here instead of this loop.

**Review by Sgitario — COMMENTED** [review]
Apart from the comments where I state a couple of minor things and a possible improvement to increase the Native coverage.
I could not run the test suite. I will revisit this PR once the dependant pull request is merged.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java
Comment (pjgg) [reply 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityTest.java:42]: For this scenario, I prefer it as it is.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:59]: I would use the official way from Quarkus to select a random port: https://quarkus.io/guides/getting-started-testing#controlling-the-test-port

**Review by Sgitario — COMMENTED** [review]
Just one comment, but I will wait to approve this PR once I can test it.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (pjgg) [reply 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:59]: **Note:**  random port only works if is defined on your `application.properties`
```
quarkus.http.test-port=0
```

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:16]: This class does not need to implement QuarkusTestResourceLifecycleManager. Also, the methods start and stop are not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:17]: This field is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:22]: This field is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:30]: This method can be removed when not using the interface.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:35]: This method can be removed when not using the interface.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityIT.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/BackwardCompatibilityIT.java:6]: Brackets is not well formatted. It should be in separated lines.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointIT.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointIT.java:6]: Same format issue here.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointNonRootPathIT.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointNonRootPathIT.java:6]: Same format issue here.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointNonRootPathTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointNonRootPathTest.java:3]: This import is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointTest.java:3]: This import is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointTestNonBaseRootPathTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/NonAppEndpointTestNonBaseRootPathTest.java:3]: This import is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointIT.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointIT.java:6]: Same format issue here.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointTest.java:3]: This import is not used.

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointTest.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/RelativePathNonAppEndpointTest.java:29]: This should not be necessary if we set `.setRun(true)` in `QuarkusProdModeTest`, however I see that for native this is not working properly. I will investigate it further and report it if needed. 
No need to change anything.

**Review by Sgitario — CHANGES_REQUESTED** [review]

File: 020-quarkus-http-non-application-endpoints/src/main/resources/application.properties
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/main/resources/application.properties:7]: This property is not doing anything, it should be removed as well.

Comment (Sgitario) [reply 020-quarkus-http-non-application-endpoints/src/main/resources/application.properties:6]: This property is still present. I meant "quarkus.http.test-port=0"

File: 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java
Comment (Sgitario) [inline 020-quarkus-http-non-application-endpoints/src/test/java/io/quarkus/qe/non_application/endpoint/CommonNonAppEndpoint.java:5]: Double empty lines here.

**Review by Sgitario — COMMENTED** [review]
Two minor comments left.

**Review by Sgitario — APPROVED** [review]

**Comment by rsvoboda** [comment]
CI failed on scenario 013
2021-03-03 11:57:21,026 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-2) HTTP Request to /rest-pong failed, error id: 320e4d93-d109-4acd-a171-dd6a877c19a8-1: io.vertx.core.VertxException: Invalid url: null

Not related to this PR, but needs some investigation .. @pjgg / @Sgitario fyi.

**Comment by Sgitario** [comment]
> CI failed on scenario 013
> 2021-03-03 11:57:21,026 ERROR [io.qua.ver.htt.run.QuarkusErrorHandler] (vert.x-eventloop-thread-2) HTTP Request to /rest-pong failed, error id: 320e4d93-d109-4acd-a171-dd6a877c19a8-1: io.vertx.core.VertxException: Invalid url: null
> 
> Not related to this PR, but needs some investigation .. @pjgg / @Sgitario fyi.

This is the cause of these failures: https://github.com/quarkusio/quarkus/pull/14953
I will ask Sergey

**Comment by rsvoboda** [comment]
Thanks, please create issue for that. I will merge this PR.

**Review by rsvoboda — APPROVED** [review]

**Pull request merged**
